libkbfs/disk_block_cache_remote.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"context"
+	"net"
+
+	"github.com/keybase/client/go/libkb"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-framed-msgpack-rpc/rpc"
+	"github.com/keybase/kbfs/kbfsblock"
+	"github.com/keybase/kbfs/kbfscrypto"
+	kbgitkbfs "github.com/keybase/kbfs/protocol/kbgitkbfs"
+	"github.com/keybase/kbfs/tlf"
+)
+
+type diskBlockCacheRemoteConfig interface {
+	logMaker
+}
+
+// DiskBlockCacheRemote implements a client to access a remote
+// DiskBlockCacheService. It implements the DiskBlockCache interface.
+type DiskBlockCacheRemote struct {
+	conn   net.Conn
+	client kbgitkbfs.DiskBlockCacheClient
+	log    traceLogger
+}
+
+var _ DiskBlockCache = (*DiskBlockCacheRemote)(nil)
+
+// NewDiskBlockCacheRemote creates a new remote disk cache client.
+func NewDiskBlockCacheRemote(kbCtx Context, config diskBlockCacheRemoteConfig) (
+	*DiskBlockCacheRemote, error) {
+	conn, xp, _, err := kbCtx.GetKBFSSocket(true)
+	if err != nil {
+		return nil, err
+	}
+	cli := rpc.NewClient(xp, KBFSErrorUnwrapper{},
+		libkb.LogTagsFromContext)
+	client := kbgitkbfs.DiskBlockCacheClient{Cli: cli}
+
+	return &DiskBlockCacheRemote{
+		conn:   conn,
+		client: client,
+		log:    traceLogger{config.MakeLogger("DBR")},
+	}, nil
+}
+
+// Get implements the DiskBlockCache interface for DiskBlockCacheRemote.
+func (dbcr *DiskBlockCacheRemote) Get(ctx context.Context, tlfID tlf.ID,
+	blockID kbfsblock.ID) (buf []byte,
+	serverHalf kbfscrypto.BlockCryptKeyServerHalf,
+	prefetchStatus PrefetchStatus, err error) {
+	dbcr.log.LazyTrace(ctx, "DiskBlockCacheRemote: Get %s", blockID)
+	defer func() {
+		dbcr.log.LazyTrace(ctx, "DiskBlockCacheRemote: Get %s done (err=%+v)", blockID, err)
+	}()
+
+	res, err := dbcr.client.GetBlock(ctx, kbgitkbfs.GetBlockArg{
+		keybase1.TLFID(tlfID.String()),
+		blockID.String(),
+	})
+	if err != nil {
+		return buf, serverHalf, prefetchStatus, err
+	}
+
+	serverHalf, err = kbfscrypto.ParseBlockCryptKeyServerHalf(res.ServerHalf)
+	if err != nil {
+		return nil, kbfscrypto.BlockCryptKeyServerHalf{}, prefetchStatus, err
+	}
+
+	return res.Buf, serverHalf, PrefetchStatus(res.PrefetchStatus), nil
```

Comment:
Hmmm. So the protocol would have an `int`, and `PrefetchStatus` controls the parsing?

Reply:
The protocol can continue to have its own enum, that's fine, I was just thinking it might be nice to have, for example, `PrefetchStatus.ToProtocol() kbgitkbfs.PrefetchStatus` and `FromProtocol(kbgitkbfs.PrefetchStatus) PrefetchStatus`.

Reply:
I see. So this is to have validation on both ends?

Reply:
Yeah, just a guard against the two copied const definitions drifting out of sync.

Reply:
I will

![Make it so](https://media.giphy.com/media/VLoN2iW8ii3wA/giphy.gif)